agent/consul/gateway_locator.go

```diff
@@ -0,0 +1,284 @@
+		srv:                    srv,
+		datacenter:             datacenter,
+		primaryDatacenter:      primaryDatacenter,
+		primaryGatewaysReadyCh: make(chan struct{}),
+	}
+}
+
+func (g *GatewayLocator) Run(stopCh <-chan struct{}) {
+	var lastFetchIndex uint64
+	retryLoopBackoff(stopCh, func() error {
+		idx, err := g.runOnce(lastFetchIndex)
+		if err != nil {
+			return err
+		}
+
+		lastFetchIndex = idx
+
+		return nil
+	}, func(err error) {
+		g.logger.Error("error tracking primary and local mesh gateways", "error", err)
+	})
+}
+
+func (g *GatewayLocator) runOnce(lastFetchIndex uint64) (uint64, error) {
+	// NOTE: we can't do RPC here because we won't have a token so we'll just
+	// mostly assume that our FSM is caught up enough to answer locally.  If
+	// this has drifted it's no different than a cache that drifts or an
+	// inconsistent read.
+	queryOpts := &structs.QueryOptions{
+		MinQueryIndex:     lastFetchIndex,
+		RequireConsistent: false,
+	}
+
+	var (
+		results   []*structs.FederationState
+		queryMeta structs.QueryMeta
+	)
+	err := g.srv.blockingQuery(
+		queryOpts,
+		&queryMeta,
+		func(ws memdb.WatchSet, state *state.Store) error {
+			// Get the existing stored version of this config that has replicated down.
+			// We could phone home to get this but that would incur extra WAN traffic
+			// when we already have enough information locally to figure it out
+			// (assuming that our replicator is still functioning).
+			idx, all, err := state.FederationStateList(ws)
+			if err != nil {
+				return err
+			}
+
+			queryMeta.Index = idx
+			results = all
+
+			return nil
+		})
+	if err != nil {
+		return 0, err
+	}
+
+	g.updateFromState(results)
+
+	return queryMeta.Index, nil
+}
+
+func (g *GatewayLocator) updateFromState(results []*structs.FederationState) {
+	var (
+		local   structs.CheckServiceNodes
+		primary structs.CheckServiceNodes
+	)
+	for _, config := range results {
+		retained := retainGateways(config.MeshGateways)
+		if config.Datacenter == g.datacenter {
+			local = retained
+		}
+		// NOT else-if because conditionals are not mutually exclusive
+		if config.Datacenter == g.primaryDatacenter {
+			primary = retained
+		}
+	}
+
+	primaryAddrs := renderGatewayAddrs(primary, true)
+	localAddrs := renderGatewayAddrs(local, false)
+
+	g.gatewaysLock.Lock()
+	defer g.gatewaysLock.Unlock()
+
+	changed := false
+	primaryReady := false
+	if !lib.StringSliceEqual(g.primaryGateways, primaryAddrs) {
+		g.primaryGateways = primaryAddrs
+		primaryReady = len(g.primaryGateways) > 0
+		changed = true
+	}
+	if !lib.StringSliceEqual(g.localGateways, localAddrs) {
+		g.localGateways = localAddrs
+		changed = true
+	}
+
+	if changed {
+		g.logger.Debug(
+			"new cached locations of mesh gateways",
+			"primary", primaryAddrs,
+			"local", localAddrs,
+		)
+	}
+
+	if primaryReady {
+		select {
+		case _, open := <-g.primaryGatewaysReadyCh:
+			if open {
+				close(g.primaryGatewaysReadyCh)
```

Comment:
Does this ever fire? If the chan is open given that nothing is ever sent to it, wont the recv block and so the default path always be taken? That makes it look like this close will never happen.

But perhaps the `_, open :=` variant changes the behavior of recv somehow that I'm missing? This is more Go trivia for my info because from tests it looks like this must work!

Reply:
Definitely a :bug:

The intent with this mechanism hooked into the primary-gateway retry-join operation is to have the gateway retry-join loop abort if we find replicated data for gateways, in the event that the retry-join component keeps looping.

This is a bit important as the retry-join-wan loop is blocked from starting to do its thing until _after_ we know we have gateways configured (to avoid useless error logging of the "you can't get there from here" kind).

I can trivially switch this out to a `sync.Once(func() { close(...) })` instead but then I have to figure out how to prove the mechanism I _wanted_ actually is working because clearly it never did.

Reply:
Thinking about this more I realized there's a recovery-type situation that we probably need to have a story around (can happen in a different PR) at some point.

1. user initializes with the `-primary-gateway` args
2. system uses fallback addresses from (1) to do the initial wanjoin
3. after wanjoin completes and the rest of the secondary bootstrapping happens we can replicate fedstates
4. now we can exclusively use the primary's fedstate list of gateways instead of the fallback.

But...what if there was some non-graceful rotation of the primary gateways that didn't give the secondaries time to replicate at least one of the new addresses in time? Then the secondary replicated fedstate information is useless AND since having it trumps the flag/config fallback data it will try to use gateways that don't exist "forever".

Probably need some sort of active process in the GatewayLocator that does some simple light RPC (status check?) to the wildcard address through each FSM and fallback gateway in the primary periodically and heuristically determine if the FSM gateways are still usable, and if not goes back to using the fallback ones until the FSM ones work again.

Reply:
https://github.com/hashicorp/consul/issues/7339